sdk/go/x/auto/stack.go

```diff
@@ -0,0 +1,772 @@
+// alter the Workspace pulumi.yaml file, and setting config on a Stack will modify the pulumi.<stack>.yaml file.
+// This is identical to the behavior of Pulumi CLI driven workspaces. Custom Workspace
+// implementations can be used to store Project and Stack settings as well as Config in a different format,
+// such as an in-memory data structure, a shared persistend SQL database, or cloud object storage. Regardless of
+// the backing Workspace implementation, the Pulumi SaaS Console will still be able to display configuration
+// applied to updates as it does with the local version of the Workspace today.
+//
+// The Automation API also provides error handling utilities to detect common cases such as concurrent update
+// conflicts:
+// 	uRes, err :=stack.Up(ctx)
+// 	if err != nil && IsConcurrentUpdateError(err) { /* retry logic here */ }
+package auto
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"runtime/debug"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/apitype"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v2/go/pulumi"
+	"github.com/pulumi/pulumi/sdk/v2/go/x/auto/optdestroy"
+	"github.com/pulumi/pulumi/sdk/v2/go/x/auto/optpreview"
+	"github.com/pulumi/pulumi/sdk/v2/go/x/auto/optrefresh"
+	"github.com/pulumi/pulumi/sdk/v2/go/x/auto/optup"
+)
+
+// Stack is an isolated, independently configurable instance of a Pulumi program.
+// Stack exposes methods for the full pulumi lifecycle (up/preview/refresh/destroy), as well as managing configuration.
+// Automation API stacks are addressed by a fully qualified stack name (fqsn) in the form "org/project/stack".
+// Multiple Stacks are commonly used to denote different phases of development
+// (such as development, staging and production) or feature branches (such as feature-x-dev, jane-feature-x-dev).
+type Stack struct {
+	workspace Workspace
+	fqsn      string
+}
+
+// FullyQualifiedStackName returns an appropriately formatted name to be used for Stack creation/selection.
+func FullyQualifiedStackName(org, project, stack string) string {
+	return fmt.Sprintf("%s/%s/%s", org, project, stack)
+}
+
+// NewStack creates a new stack using the given workspace, and fully qualified stack name (org/project/name).
+// It fails if a stack with that name already exists
+func NewStack(ctx context.Context, fqsn string, ws Workspace) (Stack, error) {
+	var s Stack
+	s = Stack{
+		workspace: ws,
+		fqsn:      fqsn,
+	}
+
+	err := ws.CreateStack(ctx, fqsn)
+	if err != nil {
+		return s, err
+	}
+
+	return s, nil
+}
+
+// SelectStack selects stack using the given workspace, and fully qualified stack name (org/project/name).
+// It returns an error if the given Stack does not exist. All LocalWorkspace operations will call SelectStack()
+// before running.
+func SelectStack(ctx context.Context, fqsn string, ws Workspace) (Stack, error) {
```

Comment:
What are some scenarios when users need to use this? Aren't they just calling commands on `Stack` directly?

Reply:
A stack is parented to a workspace, this API allows reusing the same workspace across stacks. For instance, you could imagine a mutli-geo deployment with one stack per geo:

1. Create a workspace with the program
2. Loop through geos selecting the appropriate stack and running Stack.Up()